checker/src/main/java/org/checkerframework/checker/nullness/KeyForAnnotatedTypeFactory.java

```diff
@@ -237,6 +244,88 @@ protected DependentTypesHelper createDependentTypesHelper() {
     return new KeyForDependentTypesHelper(this);
   }
 
+  /**
+   * Override to merge KeyFor annotations from Map receiver's type arguments into keySet() return
+   * type.
+   */
+  @Override
+  protected void addComputedTypeAnnotations(Tree tree, AnnotatedTypeMirror type, boolean iUseFlow) {
+    super.addComputedTypeAnnotations(tree, type, iUseFlow);
+
+    // Handle keySet() method invocations: merge KeyFor annotations from Map receiver's type
+    // arguments into the return type's type arguments
+    if (tree instanceof MethodInvocationTree) {
+      MethodInvocationTree methodInvocation = (MethodInvocationTree) tree;
+      if (TreeUtils.isMethodInvocation(methodInvocation, mapKeySet, getProcessingEnv())) {
+        if (type.getKind() == TypeKind.DECLARED) {
+          AnnotatedDeclaredType keySetReturnType = (AnnotatedDeclaredType) type;
+          ExpressionTree receiver = TreeUtils.getReceiverTree(methodInvocation);
+          if (receiver != null) {
+            AnnotatedTypeMirror receiverType = getAnnotatedType(receiver);
```

Comment:
This will recompute the type of the receiver which is potentially expensive.  Instead, if you move this code to `org.checkerframework.checker.nullness.KeyForTransfer#visitMethodInvocation` you can use the type of the receiver that dataflow stores in TransferInput.